rfcs/33-apps_own_scheduling.md

```diff
@@ -0,0 +1,239 @@
+# Feature Name: `apps_own_scheduling`
+
+## Summary
+
+All scheduling information is attached directly to systems, and ultimately owned and configurable by the central app (or hand-written `bevy_ecs` equivalent).
+To ensure that this can be done safely, plugins specify invariants about the ordering of their systems which must be respected.
+Hard ordering constraints (where systems must be both ordered and separated by a hard sync point) are introduced as an assert-style protection to guard against a serious and common form of plugin misconfiguration.
+
+## Motivation
+
+The current plugin model works well in simple cases, but completely breaks down when the central consumer (typically, an `App`) needs to configure when and if their systems run. The main difficulties are:
+
+- 3rd party plugins can have "unfixable" ordering ambiguities or interop issues with other 3rd party plugins
+- plugins can not be added to states, and run criteria cannot be added
```

Comment:
Allowing users to add new RunCriteria to a SystemSet defined in a plugin could invalidate assumptions made by the plugin author, especially when before/after constraints dont care if a system actually runs. This could break things in hard-to-identify ways, which concerns me.

Reply:
This is actually one of the main reasons why I like the system sets + can only configure based on a shared label design: plugin authors could carefully control the labels they're exporting to ensure that if A doesn't run, then B also doesn't run.

However, I think this is too implicit for my taste. Instead, I think that there's an important distinction there that we're missing with the way that the current `.before` works. 

If A is before B, and A is skipped, what should B do? The current behaviour is for B to proceed. But that's not always correct: in some cases B should be skipped if A is skipped.

Reply:
Yeah I agree that there are cases where "run B only if A ran" would be a useful tool and could help prevent certain instances of this problem. Haha we're starting to have quite the menagerie of order types: "the current soft-orders", "hard orders (side effects applied", and "only-if orders". We might need to come up with new names for these concepts / sort out the right way to express "soft order", "soft order with side effects", "only-if order", and "only-if order with side effects".